### DISCUSSION TITLE
Get rid of AWS dependency

### DISCUSSION THREAD

**Discussion by aslmx** [alexpdp7/frtcuop#7] [Ideas]
Hi, 

i saw your post on r/selfhosted and i'm happy to give this a try (to be able to recommend it or not ;)).

If you'd remove this AWS depndency  that would be good, because i could test it on Hetzner VPS that i can rent by the hour. 

I guess the AWS API is just used for the DNS records, right? I can set them up myself :)

So if that could be made optional maybe? 

thx+br

**Comment by alexpdp7** [comment]
Yes, you could create the records yourself with no issue on any provider. The VM itself can run anywhere- you just need ssh/sudo access.

I've created #8 to track this. I've looked for some abstraction layer so I can easily support multiple DNS providers, but there doesn't seem to be anything available.

The biggest sticky point is handling the Let's Encrypt DNS challenge- that's fairly annoying to do by hand, but I'll see what I can do.

**Reply by aslmx** [reply]
HI, Thanks for the quick response

> The biggest sticky point is handling the Let's Encrypt DNS challenge- that's fairly annoying to do by hand, but I'll see what I can do.

Why is that? Are you not using HTTP Challenge?

**Reply by alexpdp7** [reply]
One typical convention when doing XMPP (or email) is to use user@example.com, not user@xmpp.example.com, so you often want to use your main domain as your XMPP domain.

If you use a main domain and not a subdomain, that can be done in two ways:

One is to point example.com to your XMPP server. But for instance, then if you want to run a website on example.com, then it must be on the same server or do ugly port forwarding.

A better option is to use SRV records. Instead of pointing "the whole" example.com domain to a specific IP, you say "the XMPP server for example.com is x.x.x.x".

If you do that, then example.com will not necessarily be on your XMPP server, so HTTP Challenge is harder to do (you need to do that on your web server, then copy the certificate over, etc.).

DNS challenges have lots more possibilities, such as obtaining certificates for servers not exposed to the Internet.

It adds more complexity, but it's also more flexible.

**Reply by aslmx** [reply]
Thanks for the clarification. 
Personally I'm also doing DNS Challenge using acme.sh - but rather because i don't want to expose 443 at home for the HTTP Challenge.
Sure... people want to have a nice @example.com JID.

One way around this would be - in case people don't want to use your AWS API implementation - to have Users point you to a place where they will make sure will always be valid certificates - and maybe a reload command. Then people can use whatever method they like for getting a certificate there. (Like acme.sh with one of the many API plugins).

**Reply by alexpdp7** [reply]
Oh, that's a very good initial implementation- it's reasonable and functional, yet easy to implement. I'll consider that, thanks!

**Reply by alexpdp7** [reply]
I just pushed https://github.com/alexpdp7/frtcuop/commit/294393c4c94eb3f3f4e42e3d0e5afd142a944c2e , feel free to take it for a spin and report back!

**Comment by alexpdp7** [comment]
BTW, I pushed an improvement for this- you should be able to handle DNS and certs manually.
